src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java

```diff
@@ -0,0 +1,360 @@
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import org.jetbrains.annotations.NotNull;
+
+import com.minecolonies.api.colony.IColony;
+import com.minecolonies.coremod.colony.CitizenData;
+import com.minecolonies.coremod.colony.Colony;
+import com.minecolonies.coremod.colony.ColonyManager;
+import com.minecolonies.coremod.colony.managers.ICitizenManager;
+import com.minecolonies.coremod.commands.CommandEntryPointNew.ActionMenuHolder;
+import com.mojang.authlib.GameProfile;
+
+import net.minecraft.command.ICommandSender;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.util.math.BlockPos;
+import net.minecraftforge.common.util.FakePlayer;
+
+public enum ActionArgumentType
+{
+    PLAYER("player-expression", 0),
+    COLONY("colony-id", 0),
+    CITIZEN("citizen-id or full-name", 2),
+    COORDINATE_X("x-coordinate", 0),
+    COORDINATE_Y("y-coordinate", 0),
+    COORDINATE_Z("z-coordinate", 0),
+    BOOLEAN("boolean", 0),
+    INTEGER("integer", 0),
+    ;
+
+    public enum Is
+    {
+        REQUIRED,
+        OPTIONAL
+    }
+
+    @Nonnull private final String usageValue;
+    private final int allowedSpaceCount;
+
+    ActionArgumentType(@Nonnull final String usageValue, final int allowedSpaceCount)
+    {
+        this.usageValue = usageValue;
+        this.allowedSpaceCount = allowedSpaceCount;
+    }
+
+    public String getUsageValue()
+    {
+        return usageValue;
+    }
+
+    public int allowedSpaceCount()
+    {
+        return allowedSpaceCount;
+    }
+
+    private List<String> getOnlinePlayerNames(@NotNull final MinecraftServer server)
+    {
+        final String[] onlinePlayerNames = server.getOnlinePlayerNames();
+        return Arrays.asList(onlinePlayerNames);
+    }
+
+    private List<String> getColonyIdStrings()
+    {
+        final List<Colony> colonyList = ColonyManager.getColonies();
+        final List<String> colonyIdList = new ArrayList<>(colonyList.size());
+        for (final Colony colony : colonyList)
+        {
+            colonyIdList.add(String.valueOf(colony.getID()));
+        }
+        return colonyIdList;
+    }
+
+    private List<String> getCitizenNames()
+    {
+        final List<Colony> colonyList = ColonyManager.getColonies();
+        final List<String> citizenNameList = new ArrayList<>();
+        for (final Colony colony : colonyList)
+        {
+            final List<CitizenData> citizenDataList = colony.getCitizenManager().getCitizens();
+            for (final CitizenData citizenData : citizenDataList)
+            {
+                citizenNameList.add(citizenData.getName());
+            }
+        }
+        return citizenNameList;
+    }
+
+    public List<String> getTabCompletions(@NotNull final MinecraftServer server,
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This method has 127 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS138)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 18, down to the maximum allowed 6. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "getTabCompletions" is 27 which is greater than 15 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 57 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)